### PR TITLE
Fixed the SHM segment smashing while reading with security <2.1.x>

### PR DESCRIPTION
**Pull request by rsanchez15**
This is a port of PR #1644 from 2.0.x to 2.1.x

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3336/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/544/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2993/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3368/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3366/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/568/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3014/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3398/)

File: src/cpp/rtps/messages/MessageReceiver.cpp [hunk 1]
```diff
@@ -54,6 +54,7 @@ MessageReceiver::MessageReceiver(
 #if HAVE_SECURITY
     , crypto_msg_(participant->is_secure() ? rec_buffer_size : 0)
     , crypto_payload_(participant->is_secure() ? rec_buffer_size : 0)
+    , crypto_submsg_(participant->is_secure() ? rec_buffer_size : 0)
```
Comment (MiguelCompany) [inline src/cpp/rtps/messages/MessageReceiver.cpp:57]:
```suggestion
    , crypto_submsg_(participant->is_secure() ? rec_buffer_size : 0)
    , crypto_payload_(participant->is_secure() ? rec_buffer_size : 0)
```

File: include/fastdds/rtps/messages/MessageReceiver.h [hunk 2]
```diff
@@ -90,7 +90,10 @@ class MessageReceiver
     Time_t timestamp_;
 
 #if HAVE_SECURITY
+    //!Buffer to process the decoded RTPS message
     CDRMessage_t crypto_msg_;
+    //!Buffer to process each decoded RTPS sub-message
+    CDRMessage_t crypto_submsg_;
     SerializedPayload_t crypto_payload_;
```
Comment (MiguelCompany) [inline include/fastdds/rtps/messages/MessageReceiver.h:97]:
```suggestion
    CDRMessage_t crypto_submsg_;
    //!Buffer to process a decoded payload
    SerializedPayload_t crypto_payload_;
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3377/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/577/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3029/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3409/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3389/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/587/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3039/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3421/)

**Review by MiguelCompany — APPROVED** [review]

**Comment by EduPonz** [comment]
@MiguelCompany what do you think?

**Comment by MiguelCompany** [comment]
> what do you think?

I think we can merge this one

**Pull request merged**